### PR TITLE
Issues/introduce ipaddress

### PR DESCRIPTION
**Pull request by AndreyDegtyarev**
Introduced IpAddress package
Created Readme.md for IpAddress package
Added reference in Nox.Demo

### CODE REVIEW

File: src/Nox.Reference.DataLoaders/Program.cs [hunk 1]
```diff
@@ -12,10 +13,12 @@
 
         WorldDbContext.UseDatabaseConnectionString(config.GetConnectionString(ConfigurationConstants.WorldDataLoadConnectionStringName)!);
         MachineDbContext.UseDatabaseConnectionString(config.GetConnectionString(ConfigurationConstants.MachineDataLoadConnectionStringName)!);
+        IpAddressContext.UseDatabaseConnectionString(config.GetConnectionString(ConfigurationConstants.IpAddressConnectionStringName)!);
 
         services.AddScoped<DataSeedRunner>();
-        services.AddMachineContext();
-        services.AddWorldContext();
+        //services.AddMachineContext();
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.DataLoaders/Program.cs:19]: why

File: tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        [TestCase("142.251.32.46", "US")]
+        [TestCase("149.126.4.31", "CH")]
+        [TestCase("2a01:ab20:0:4::31", "CH")]
+        [TestCase("2001:470:71:21::", "UA")]
+        [TestCase("2e00::", "DE")]
+        [TestCase("2607:f8b0:3d00::", "US")]
+        [TestCase("2001:4860:4860::", "US")]
+        public void IpAddressStatic_DetermineCountryByIpCode_Success(string ipAddress, string expectedCountryCode)
+        {
+            var result = IpAddressContext.GetCountryByIp(ipAddress);
```
Comment (ILoveMyTortoise) [inline tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs:61]: why IpAddressContext and not just 'IpAddress'?

File: tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        [TestCase("2607:f8b0:3d00::", "US")]
+        [TestCase("2001:4860:4860::", "US")]
+        public void IpAddress_DetermineCountryByIpCode_Success(string ipAddress, string expectedCountryCode)
+        {
+            var result = _ipAddressService.GetCountryByIp(ipAddress);
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(result.Kind, Is.EqualTo(IpSearchResultKind.Success));
+                Assert.That(expectedCountryCode, Is.EqualTo(result.CountryCode));
```
Comment (ILoveMyTortoise) [inline tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs:43]: Hm
maybe we should join it with world context, and return why result.Country.Code instead. May need to discuss

File: src/Nox.Reference.Data.World/Nox.Reference.Data.World.csproj [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
   <PropertyGroup>
     <TargetFramework>net6.0</TargetFramework>
     <ImplicitUsings>enable</ImplicitUsings>
     <Nullable>enable</Nullable>
     <PackageId>Nox.Reference.Data.World</PackageId>
-    <AssemblyVersion>1.0.9.0</AssemblyVersion>
-    <FileVersion>1.0.9.0</FileVersion>
-    <PackageVersion>1.0.9.0</PackageVersion>
+    <AssemblyVersion>1.0.0.9</AssemblyVersion>
+    <FileVersion>1.0.0.9</FileVersion>
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.World/Nox.Reference.Data.World.csproj:8]: why? I don't think we should do it, leave 1.0.9 or use 1.0.10

File: src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
...^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$))");
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs:12]: add some comment, and move to a different file, just so it doesn't take space
Ideally we want also to split by lines

File: src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    private readonly IpAddressDbContext _dbContext;
+
+    /// <summary>
+    /// IpAddressService constructor.
+    /// </summary>
+    /// <param name="dbContext"></param>
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs:18]: let's not leave empty docs

File: src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        _dbContext = dbContext;
+    }
+
+    /// <summary>
+    /// Get country code for ip address.
+    /// </summary>
+    /// <param name="ipAddress">Ipv4 or Ipv6 address string</param>
+    /// <returns>Two letter Country Code</returns>
+    /// <exception cref="ApplicationException"></exception>
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.IpAddress/Services/IpAddressService.cs:29]: here too

File: src/Nox.Reference.Data.IpAddress/README.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+Run command
+
+`
+dotnet ef database update --connection "Data Source=..\\..\\data\\Nox.Reference.Data.IpAddress.db"
+`
+
+Nox.Reference.IpAddress.db database file will appear in (RootPath)\data\output\sqlite
+
+
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.IpAddress/README.md:101]: why empty lines

File: src/Nox.Reference.Data.IpAddress/DataContext/IpAddressDbContext.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        var connectionString = _databasePath ?? _configuration.GetConnectionString(ConfigurationConstants.IpAddressConnectionStringName);
+
+        // TODO: fix adding migrations. Currently throws an error of "empty db path". Need to find a way of fixing it.
+        connectionString = DatabasePathHelper.FixConnectionStringPathUsingAssemblyPath(connectionString, typeof(IpAddressDbContext), nameof(IpAddressContext));
+
+        optionsBuilder.UseSqlite(connectionString);
+
+        // optionsBuilder
+        //.UseLazyLoadingProxies()
```
Comment (ILoveMyTortoise) [inline src/Nox.Reference.Data.IpAddress/DataContext/IpAddressDbContext.cs:58]: why

File: src/Nox.Demo/IpAddressDemo.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            var services = new ServiceCollection();
+            services.AddIpAddressContext();
+
+            _ipAddressService = services.BuildServiceProvider().GetRequiredService<IIpAddressService>();
+        }
+
+        public static void ShowDemo()
+        {
+            var ipAddressExamples = new List<(string ip, string country)>
```
Comment (ILoveMyTortoise) [inline src/Nox.Demo/IpAddressDemo.cs:25]: let's leave less for demo, just 5 at most

File: tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs (hunk 2 above)
Comment (AndreyDegtyarev) [reply tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs:61]: In order not to overlap IpAddress entity name. Btw, IpAddressInfo is reserved to dto.
Any suggestion?

File: src/Nox.Reference.Data.World/Nox.Reference.Data.World.csproj (hunk 4 above)
Comment (AndreyDegtyarev) [reply src/Nox.Reference.Data.World/Nox.Reference.Data.World.csproj:8]: Set 1.0.0.9

File: tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs (hunk 2 above)
Comment (ILoveMyTortoise) [reply tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs]: Hm
I think overlapping is fine
Ideally -- discuss with Alex and Andre
Have no strong opinion here

File: tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs (hunk 3 above)
Comment (ILoveMyTortoise) [reply tests/Nox.Reference.Data.IpAddress.Tests/IpAddress.Tests.cs:45]: weird, but ok

**Comment by ILoveMyTortoise** [comment]
Just one comment, about docs
I saw removed exception docs, but better to have them and describe situation, on which this exception can happen
Also, let's not have just <returns>SomeType<returns>, but better some human-readable description, like <returns>IPAddress detection result object with error or country code<returns>, just as en example

**Comment by AndreyDegtyarev** [comment]
> Just one comment, about docs I saw removed exception docs, but better to have them and describe situation, on which this exception can happen Also, let's not have just SomeType, but better some human-readable description, like IPAddress detection result object with error or country code, just as en example

Logic was changed. It doesn't throw exception at all. It should get Search Result (Success,InvalidInput, NotFound)

**Comment by AndreyDegtyarev** [comment]
> Just one comment, about docs I saw removed exception docs, but better to have them and describe situation, on which this exception can happen Also, let's not have just SomeType, but better some human-readable description, like IPAddress detection result object with error or country code, just as en example

I guess these are straight forward results. What else can be explained if IpAddress not found, or InvalidInput. Surely, it's only single property with Message, but it's extra property.

**Review by ILoveMyTortoise — APPROVED** [review]

**Pull request merged**